runtime/src/accounts_db.rs

```diff
@@ -1397,8 +1426,52 @@ impl AccountsDB {
         hashes
     }
 
+    pub fn freeze_accounts(&self, ancestors: &HashMap<Slot, usize>, account_pubkeys: &[Pubkey]) {
+        let mut frozen_accounts = self.frozen_accounts.lock().unwrap();
+        for account_pubkey in account_pubkeys {
+            let (account, _slot) = self
+                .load_slow(ancestors, &account_pubkey)
+                .unwrap_or_default();
+            let frozen_account_info = FrozenAccountInfo {
+                hash: Self::hash_frozen_account_data(&account),
+                lamports: account.lamports,
+            };
+            warn!(
+                "Account {} is now frozen at lamports={}, hash={}",
+                account_pubkey, frozen_account_info.lamports, frozen_account_info.hash
+            );
+            frozen_accounts.insert(*account_pubkey, frozen_account_info);
+        }
+    }
+
     /// Store the account update.
     pub fn store(&self, slot_id: Slot, accounts: &[(&Pubkey, &Account)]) {
+        let mut frozen_accounts = self.frozen_accounts.lock().unwrap();
```

Comment:
mutex here is very bad for perf

Reply:
Ah, thanks!  I guess I'm plumbing this through `AccountsDb::new()` then unless you see a different way?

Reply:
Well. If you can get rid of locks entirely that is best, either by plumbing through `new` or having a `set` function that can set it after the fact.

2nd best option is RwLock, and you'll want to structure it so you are doing the least amount of work in the write critical section that you need here and ideally you have some other indication that frozen_accounts is empty outside of the lock so you don't need to take the lock to check. I'm not sure this extra hashing is required, but if it is, then do the hashing outside the lock, take the the read lock to check the values and keep track if any need updating. Then drop read lock, take write lock if necessary and then update the values if necessary. AccountsDB:update_index is kind of an example of this.

One idea is to maybe store this frozen flag in the `AccountsIndex` itself, maybe uses more ram though.

Reply:
> If you can get rid of locks entirely that is best, either by plumbing through new or having a set function that can set it after the fact.

I'm doing this now.  It doesn't look too bad

Reply:
> It doesn't look too bad
🤕